exonum/src/blockchain/config.rs

```diff
@@ -295,6 +296,130 @@ impl ValidateInput for ConsensusConfig {
     }
 }
 
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    Serialize,
+    Deserialize,
+    ProtobufConvert,
+    BinaryValue,
+    ObjectHash,
+)]
+#[protobuf_convert(source = "runtime::GenesisConfig")]
+pub struct GenesisConfig {
+    /// Blockchain configuration used to create the genesis block.
+    pub consensus_config: ConsensusConfig,
+
+    /// Artifacts specification of builtin services.
+    pub artifacts: Vec<ArtifactSpec>, // TODO: HashSet? Duplications? Perhaps check in builder?
+
+    /// List of the privileged services with the configuration parameters that are created directly
+    /// in the genesis block.
+    pub builtin_instances: Vec<ConfiguredInstanceSpec>,
+}
+
+/// Represents data that is required for initialization of service instance.
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    Serialize,
+    Deserialize,
+    ProtobufConvert,
+    BinaryValue,
+    ObjectHash,
+)]
+#[protobuf_convert(source = "runtime::ConfiguredInstanceSpec")]
+// TODO: Think about better name and/or substitution the `InstanceSpec`
+pub struct ConfiguredInstanceSpec {
+    /// Wrapped `InstanceSpec`.
+    pub instance_spec: InstanceSpec,
+    /// Constructor argument for specific `InstanceSpec`.
+    pub constructor: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct GenesisConfigBuilder {
+    /// Consensus config.
+    consensus_config: ConsensusConfig,
+    /// Artifacts specifications for builtin services.
+    artifacts: Vec<ArtifactSpec>,
```

Comment:
It could make sense to verify (or ensure) that `artifacts` are unique, e.g., by using `HashMap<ArtifactId, Vec<u8>>` here. Similarly, it could make sense to verify uniqueness of instances during the building process.